pkg/operator/environmentchecker/environmentcheckcontroller.go

```diff
@@ -0,0 +1,409 @@
+	nodeLister := apiClients.NodeInformer.Lister()
+
+	rc := recorder.WithComponentSuffix("vmware-" + strings.ToLower(name))
+
+	c := &EnvironmentCheckController{
+		name:            name,
+		targetNamespace: targetNamespace,
+		kubeClient:      apiClients.KubeClient,
+		operatorClient:  apiClients.OperatorClient,
+		configMapLister: configMapInformer.Lister(),
+		secretLister:    apiClients.SecretInformer.Lister(),
+		csiNodeLister:   csiNodeLister,
+		scLister:        scInformer.Lister(),
+		csiDriverLister: csiDriverLister,
+		nodeLister:      nodeLister,
+		apiClients:      apiClients,
+		eventRecorder:   rc,
+		vSphereChecker:  newVSphereEnvironmentChecker(),
+		infraLister:     infraInformer.Lister(),
+	}
+	c.controllers = []conditionalController{}
+	c.createCSIDriver()
+	c.storageClassController = c.createStorageClassController()
+
+	return factory.New().WithInformers(
+		apiClients.OperatorClient.Informer(),
+		configMapInformer.Informer(),
+		apiClients.SecretInformer.Informer(),
+		infraInformer.Informer(),
+		scInformer.Informer(),
+	).WithSync(c.sync).
+		ResyncEvery(resyncDuration).
+		WithSyncDegradedOnError(apiClients.OperatorClient).ToController(c.name, rc)
+}
+
+func (c *EnvironmentCheckController) sync(ctx context.Context, syncContext factory.SyncContext) error {
+	klog.V(4).Infof("%s: sync started", c.name)
+	defer klog.V(4).Infof("%s: sync complete", c.name)
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+
+	if opSpec.ManagementState != operatorapi.Managed {
+		klog.V(4).Infof("%s: ManagementState is %s, skipping", c.name, opSpec.ManagementState)
+		return nil
+	}
+
+	infra, err := c.infraLister.Get(infraGlobalName)
+	if err != nil {
+		return err
+	}
+
+	if infra.Status.PlatformStatus == nil {
+		klog.V(4).Infof("Unknown platform: infrastructure status.platformStatus is nil")
+		return nil
+	}
+
+	if infra.Status.PlatformStatus.Type != ocpv1.VSpherePlatformType {
+		klog.V(4).Infof("Unsupported platform: infrastructure status.platformStatus.type is %s", infra.Status.PlatformStatus.Type)
+		return nil
+	}
+
+	delay, result, checkRan, ocpError := c.runClusterCheck(ctx, infra)
+	if ocpError != nil {
+		return ocpError
+	}
+
+	// if checks did not run
+	if !checkRan {
+		return nil
+	}
+	queue := syncContext.Queue()
+	queueKey := syncContext.QueueKey()
+
+	klog.V(2).Infof("Scheduled the next check in %s", delay)
+	time.AfterFunc(delay, func() {
+		queue.Add(queueKey)
+	})
+
+	if c.shouldDegradeCluster(result) {
+		return result.CheckError
+	}
+
+	// if operand was not started previously and block upgrade is false and clusterdegrade is also false
+	// then and only then we should start CSI driver operator
+	if !c.operandControllerStarted && !result.BlockUpgrade && !result.ClusterDegrade {
+		go c.runConditionalController(ctx)
+		c.operandControllerStarted = true
+	}
+
+	// sync storage class
+	storageClassSyncResult := c.storageClassController.Sync(ctx, c.vSphereConnection)
```

Comment:
I think we shouldn't sync the controller in here, it should run on its own goroutine. For example, if the StorageClass is deleted, only the StorageClass controller should re-sync.

Reply:
hmm, it is tricky to share vsphere connection between parallel controllers. We will need a global connection pool with caching and everything.